### ISSUE TITLE
Config file order lines changed every puppet run

### ISSUE THREAD

**Issue by mrh666** [fraenki/puppet-proftpd#12]
Hi there,

Every puppet agent run I've got the same config line with a different lines or blocks order, e.g.:

```
-ControlsEngine on
 AuthOrder mod_auth_pam.c* mod_auth_unix.c
-ServerName FTP server
+ControlsEngine on
 TransferLog /var/log/proftpd/xferlog
-ServerAdmin root@31.16.135
+ServerName FTP server
 DefaultRoot ~ !adm
+ServerAdmin root@31.16.135
 LogFormat default  "%h %l %u %t \"%r\" %s %b"
 LogFormat auth     "%v [%P] %h %t \"%r\" %s"
 SystemLog /var/log/proftpd/proftpd.log
+DefaultServer on
 ModuleControlsACLs insmod,rmmod allow user root
 ModuleControlsACLs lsmod allow user *
-UseReverseDNS off
 UseSendfile off
-DefaultServer on
+UseReverseDNS off
 Group nobody
 AuthPAMConfig proftpd
 ServerIdent on "FTP Server ready."
 ControlsACLs all allow user root
-User nobody
 MaxInstances 10
 ControlsSocketACL allow user *
+User nobody
 ControlsLog /var/log/proftpd/controls.log
```

and next run:

```
+<Directory /test>
+  <Limit ALL>
+    AllowAll true
+  </Limit>
+</Directory>
+
+<IfModule mod_vroot.c>
+  VRootEngine on
+</IfModule>
+
 <Limit LOGIN>
   AllowUser ftp
 </Limit>

 <IfModule mod_ctrls_admin.c>
-  AdminControlsACLs all allow user root
   AdminControlsEngine on
+  AdminControlsACLs all allow user root
 </IfModule>

-<Directory /test>
-  <Limit ALL>
-    AllowAll true
-  </Limit>
-</Directory>
-
 <IfModule mod_tls_shmcache.c>
   TLSSessionCache shm:/file=/var/run/proftpd/sesscache
 </IfModule>

-AuthOrder mod_auth_pam.c* mod_auth_unix.c
 ControlsEngine on
-TransferLog /var/log/proftpd/xferlog
-ServerName FTP server
-DefaultRoot ~ !adm
-ServerAdmin root@31.16.135
-LogFormat default  "%h %l %u %t \"%r\" %s %b"
-LogFormat auth     "%v [%P] %h %t \"%r\" %s"
-SystemLog /var/log/proftpd/proftpd.log
+ControlsSocketACL allow user *
+User nobody
+ControlsLog /var/log/proftpd/controls.log
+AuthPAMConfig proftpd
+AuthOrder mod_auth_pam.c* mod_auth_unix.c
 DefaultServer on
+Group nobody
 ModuleControlsACLs insmod,rmmod allow user root
 ModuleControlsACLs lsmod allow user *
 UseSendfile off
+MaxInstances 10
+ServerName FTP server
+ControlsACLs all allow user root
+SystemLog /var/log/proftpd/proftpd.log
+ServerAdmin root@31.16.135
 UseReverseDNS off
-Group nobody
-AuthPAMConfig proftpd
 ServerIdent on "FTP Server ready."
-ControlsACLs all allow user root
-MaxInstances 10
-ControlsSocketACL allow user *
-User nobody
-ControlsLog /var/log/proftpd/controls.log
+LogFormat default  "%h %l %u %t \"%r\" %s %b"
+LogFormat auth     "%v [%P] %h %t \"%r\" %s"
+TransferLog /var/log/proftpd/xferlog
+DefaultRoot ~ !adm

 <Global>
-  AllowOverwrite yes
-  Umask 022
   RequireValidShell off
+  Umask 022
+  AllowOverwrite yes
   <Limit ALL SITE_CHMOD>
     AllowAll
   </Limit>
 </Global>

-<IfModule mod_vroot.c>
-  VRootEngine on
-</IfModule>
```

And it's triggers proftpd restart. 
Any idea how to sort it once and for all?

**Comment by fraenki** [comment]
@mrh666: Please paste your proftp manifest (or hiera entries) here. Which version of puppet are you running?

**Comment by mrh666** [comment]
Module version: fraenki-proftpd (v1.4.2)

Puppet agent: 3.8.7

Puppet master: 3.7.3

init.pp

```
class ftpserver(
  $consul_service_name = 'ftpserver'
) {

  file { [ '/var/ftp', '/var/ftp/test' ]:
    ensure => directory,
    owner  => "ftp",
    group  => "ftp",
    mode   => 755,
  } ->

  class { 'proftpd':
    options => {
      'ROOT'  => {
        'ServerName'   => 'FTP server',
        'MaxInstances' => '10',
      },
      'IfModule mod_vroot.c' => {
        'VRootEngine' => 'on',
      },
      'Limit LOGIN' => {
        'AllowUser' => 'ftp',
      },
      'Directory /test' => {
        'Limit ALL' => {
          'AllowAll' => 'true',
        }
      }
    },
  }

  file {'/etc/pam.d/proftpd':
    ensure => present,
    owner  => $user,
    group  => $group,
    mode   => 0644,
    source => "puppet:///modules/${module_name}/pam.proftpd",
    notify  => Service['proftpd'],
  }

  consulagent::service { $consul_service_name:
    tags    => [$consul_service_name],
  }

  user { 'ftp':
    ensure   => present,
    password => '$1$9u77777777777771mul6ELT0Fl.',
  }

}
```

pam.proftpd:

```
#%PAM-1.0M-1.0
auth required pam_listfile.so item=user sense=deny file=/etc/ftpusers onerr=succeed
auth required pam_shells.so
auth include system-auth
account include system-auth
#session include system-auth
session required pam_loginuid.so
```

**Comment by nomoresecrets** [comment]
same issue here, conf file is being recreated every puppet run

**Comment by fraenki** [comment]
I can't reproduce this behaviour with the configuration provided by @mrh666.
Please paste the output of your puppet run here (`puppet agent -t -v`).

**Comment by fraenki** [comment]
https://tickets.puppetlabs.com/browse/PUP-1755 suggests that this is related to Puppet on Ruby 1.8. Are you guys using Ruby 1.8? I've tested Ruby 2.0 (CentOS 7) and Ruby 1.9 (Ubuntu 14.04).

**Comment by fraenki** [comment]
I've added this as a known issue to the README. I'm sorry, there's no fix for Ruby 1.8 (because I can't sort the config hash without breaking the config).
